data/transactions/logic/eval.go

```diff
@@ -4060,3 +4062,126 @@ func opBase64Decode(cx *EvalContext) {
 	}
 	cx.stack[last].Bytes, cx.err = base64Decode(cx.stack[last].Bytes, encoding)
 }
+func hasDuplicateKeys(dec *json.Decoder) (bool, error) {
+	keys := make(map[string]bool)
+	var value interface{}
+	dec.Token() // {
+	for dec.More() {
+		// Get JSON key
+		key, err := dec.Token()
+		if err != nil {
+			return false, err
+		}
+		// end of json
+		if key == '}' {
+			break
+		}
+		// decode value
+		err = dec.Decode(&value)
+		if err != nil {
+			return false, err
+		}
+		// check for duplicates
+		if _, ok := keys[key.(string)]; ok {
+			return true, nil
+		}
+		keys[key.(string)] = true
+	}
+	return false, nil
+}
+
+func parseJSON(jsonText []byte) (interface{}, error) {
+	if !json.Valid(jsonText) {
+		return nil, fmt.Errorf("invalid json text")
+	}
+	// check for duplicate keys
+	decoder := json.NewDecoder(bytes.NewReader(jsonText))
+	hasDuplicates, err := hasDuplicateKeys(decoder)
+	if hasDuplicates {
+		return nil, fmt.Errorf("invalid json text, duplicate keys not allowed")
+	}
+	// decode text
+	decoder = json.NewDecoder(bytes.NewReader(jsonText))
+	decoder.UseNumber()
+	var parsed interface{}
+	err = decoder.Decode(&parsed)
+	if err != nil {
+		return nil, err
+	}
+	return parsed, nil
+}
+func opJSONRef(cx *EvalContext) {
+	// get json key
+	last := len(cx.stack) - 1
+	key := string(cx.stack[last].Bytes)
+	cx.stack = cx.stack[:last] // pop
+
+	//parse json text
+	last = len(cx.stack) - 1
+	resp, err := parseJSON(cx.stack[last].Bytes)
+	if err != nil {
+		cx.err = fmt.Errorf("error while parsing JSON text, %v", err)
+		return
+	}
+
+	//get value from json
+	var stval stackValue
+	resp, ok := resp.(map[string]interface{})[key]
+	if !ok {
+		cx.err = fmt.Errorf("key %s not found in JSON text", key)
+		return
+	}
+	expectedType := jsonRefTypeNames[cx.program[cx.pc+1]]
+	switch expectedType {
+	case "JSONString":
+		val, ok := resp.(string)
+		if !ok {
+			cx.err = fmt.Errorf("got type %T but expected string", resp)
+			return
+		}
+		stval.Bytes = []byte(val)
+	case "JSONUint64":
+		val, ok := resp.(json.Number)
+		if !ok {
+			cx.err = fmt.Errorf("got type %T but expected uint64", resp)
+			return
+		}
+		// val is not a float
+		if strings.Contains(val.String(), ".") {
+			cx.err = fmt.Errorf("got type float64 but expected uint64")
+			return
+		}
+		// val is not negative
+		if val.String()[0] == '-' {
+			cx.err = fmt.Errorf("JSON value should be a uint64")
+			return
+		}
+		// val > max uint64
+		maxVal := strconv.FormatUint(math.MaxUint64, 10)
+		if val.String() > maxVal {
+			cx.err = fmt.Errorf("JSON value range within uint64 range")
+			return
+		}
+		// convert to uint64
+		var bigInt big.Int
+		bigInt.UnmarshalText([]byte(val.String()))
+		stval.Uint = bigInt.Uint64()
```

Comment:
Perhaps it's better to do the conversion this way (and detect failures - I assume that's possible?) rather than try to check the string before conversion.  For example I think "1e200" is a valid json number (and it's a float, sort of), but it's bigger that uint64, despite the string check above.

`strconv.ParseUint(val.String(), 10, 64)` might be all we need.

Though if we do that, we need to declare that 1e3 is _not_ acceptable.  Though I'm thinking maybe it should be allowed?  Do people think of that as an int or not?

Reply:
+1 to using `strconv.ParseUint`, since `big.Int`'s `UnmarshalText` also parses hex and binary